### PR TITLE
Add nullable parameters

### LINKED ISSUE #109: Change placeholder to nullable

### PR DESCRIPTION
**Pull request by kzfsg**
Fixes #109

### CODE REVIEW

File: src/main/java/seedu/address/model/company/Company.java [hunk 1]
```diff
@@ -12,7 +12,8 @@
 
 /**
  * Represents a Company in the address book.
- * Guarantees: details are present and not null, field values are validated, immutable.
+ * Guarantees: name, tags, remark, and status are not null; phone, email, and address can be null;
```
Comment (jeffreylim24) [inline src/main/java/seedu/address/model/company/Company.java]: It says remark cannot be null here, but later on in Remark.java it says remark can be null.

File: src/main/java/seedu/address/logic/Messages.java [hunk 2]
```diff
@@ -38,22 +38,27 @@ public static String format(Company company) {
...
-        String emailDisplay = company.getEmail().value.equals("noemailprovided@placeholder.com")
+        String emailDisplay = company.getEmail().value == null
                 ? "No email provided"
                 : company.getEmail().value;
 
+        // Format address with user-friendly placeholder
+        String addressDisplay = company.getAddress().value == null
+                ? "No address provided"
+                : company.getAddress().value;
+
```
Comment (jeffreylim24) [inline src/main/java/seedu/address/logic/Messages.java]: Now that we are allowing address to be null, adding this makes sense. As for remark, I realise this method here doesn't have remark included. Should we also add one for remark? and If remark is a null value instead of empty string maybe we can use the same logic for address to display Remark.

File: src/main/java/seedu/address/storage/JsonAdaptedCompany.java
Comment (jeffreylim24) [inline src/main/java/seedu/address/storage/JsonAdaptedCompany.java]: same issue for remark as mentioned above

File: src/test/java/seedu/address/logic/parser/EditCommandParserTest.java [hunk 3]
```diff
@@ -95,7 +94,7 @@ public void parse_invalidValue_failure() {
         assertParseFailure(parser, "1" + INVALID_NAME_DESC, Name.MESSAGE_CONSTRAINTS); // invalid name
         assertParseFailure(parser, "1" + INVALID_PHONE_DESC, Phone.MESSAGE_CONSTRAINTS); // invalid phone
         assertParseFailure(parser, "1" + INVALID_EMAIL_DESC, Email.MESSAGE_CONSTRAINTS); // invalid email
-        assertParseFailure(parser, "1" + INVALID_ADDRESS_DESC, Address.MESSAGE_CONSTRAINTS); // invalid address
+        // Note: blank address (INVALID_ADDRESS_DESC) is now valid as it clears the address field
```
[the remark this answers is not in the document]
Comment (kzfsg) [reply src/test/java/seedu/address/logic/parser/EditCommandParserTest.java]: Added

File: src/main/java/seedu/address/model/company/Address.java [hunk 4]
```diff
@@ -22,19 +22,20 @@ public class Address {
...
+            checkArgument(isValidAddress(address), MESSAGE_CONSTRAINTS);
+        }
         value = address;
     }
 
     /**
-     * Returns true if a given string is a valid email.
+     * Returns true if a given string is a valid address.
      */
     public static boolean isValidAddress(String test) {
```
Comment (soakedumplings) [inline src/main/java/seedu/address/model/company/Address.java]: Should we rename `test` to `address` to better reflect its purpose? This could improve code readability and maintainability. Just some nits, both works fine

File: src/main/java/seedu/address/model/company/Phone.java [hunk 5]
```diff
@@ -18,19 +18,20 @@ public class Phone {
...
+        }
         value = phone;
     }
 
     /**
      * Returns true if a given string is a valid phone number.
      */
     public static boolean isValidPhone(String test) {
-        return test.matches(VALIDATION_REGEX);
+        return test != null && test.matches(VALIDATION_REGEX);
```
Comment (soakedumplings) [inline src/main/java/seedu/address/model/company/Phone.java]: same variable naming issue as above

File: src/main/java/seedu/address/storage/JsonAdaptedCompany.java
Comment (soakedumplings) [inline src/main/java/seedu/address/storage/JsonAdaptedCompany.java]: Should change this and do what you did above for phone, email and address as well since now we allow remark to be null

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 6]
```diff
@@ -137,8 +136,8 @@ public void parse_repeatedNonTagValue_failure() {
         assertParseFailure(parser, validExpectedCompanyString + INVALID_PHONE_DESC,
                 Messages.getErrorMessageForDuplicatePrefixes(PREFIX_PHONE));
 
-        // invalid address
-        assertParseFailure(parser, validExpectedCompanyString + INVALID_ADDRESS_DESC,
+        // valid address (blank addresses are now allowed, but duplicates are not)
+        assertParseFailure(parser, validExpectedCompanyString + VALID_ADDRESS_DESC,
```
Comment (nathantew14) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: this test also tests that when parsing an add command with an invalid value and a duplicate parameter, the duplicate parameter triggers the error rather than the invalid value; since address has no possible invalid values, this test should be removed

File: src/test/java/seedu/address/logic/commands/CommandTestUtil.java [hunk 7]
```diff
@@ -61,7 +61,7 @@ public class CommandTestUtil {
     public static final String INVALID_NAME_DESC = " " + PREFIX_NAME + "James&"; // '&' not allowed in names
     public static final String INVALID_PHONE_DESC = " " + PREFIX_PHONE + "911a"; // 'a' not allowed in phones
     public static final String INVALID_EMAIL_DESC = " " + PREFIX_EMAIL + "bob!yahoo"; // missing '@' symbol
-    public static final String INVALID_ADDRESS_DESC = " " + PREFIX_ADDRESS; // empty string not allowed for addresses
+    public static final String VALID_ADDRESS_DESC = " " + PREFIX_ADDRESS + " "; // empty address; technically valid
```
Comment (nathantew14) [inline src/test/java/seedu/address/logic/commands/CommandTestUtil.java]: doesn't make sense for VALID_ADDRESS_DESC to be here; since there are no possible invalid addresses, just remove this line

**Review by nathantew14 — CHANGES_REQUESTED** [review]

**Review by soakedumplings — COMMENTED** [review]
Some comments with regards to accepting null for remark. Limit Remark to null and non empty string only so that it behaves the same way as email, phone and address for cohesiveness of product.


Others are some nits.

**Review by nathantew14 — CHANGES_REQUESTED** [review]

File: src/main/java/seedu/address/logic/parser/AddCommandParser.java [hunk 8]
```diff
@@ -51,21 +51,21 @@ public AddCommand parse(String args) throws ParseException {
...
-            email = new Email("noemailprovided@placeholder.com");
+            email = new Email(null);
         }
 
         Address address;
         if (argMultimap.getValue(PREFIX_ADDRESS).isPresent()) {
             address = ParserUtil.parseAddress(argMultimap.getValue(PREFIX_ADDRESS).get());
         } else {
-            address = new Address("No address provided");
+            address = new Address(null);
```
[the remark this answers is not in the document]
Comment (nathantew14) [reply src/main/java/seedu/address/logic/parser/AddCommandParser.java]: will document in DG as stated in #169

File: src/main/java/seedu/address/model/company/Company.java [hunk 9]
```diff
@@ -28,11 +29,15 @@ public class Company {
...
+     * Name, tags and status must be present and not null.
+     * Phone, email, remark and address can be null.
      */
     public Company(Name name, Phone phone, Email email, Address address, Set<Tag> tags, Remark remark,
                    Status status) {
-        requireAllNonNull(name, phone, email, address, tags, remark, status);
+        requireNonNull(name);
+        requireNonNull(tags);
+        requireNonNull(remark);
+        requireNonNull(status);
```
Comment (nathantew14) [inline src/main/java/seedu/address/model/company/Company.java]: all fields should still be non-null as "empty" Phone, Address objects holding null should be passed in instead

File: src/main/java/seedu/address/model/company/Company.java [hunk 10]
```diff
@@ -28,11 +29,15 @@ public class Company {
...
         this.name = name;
```
[the remark this answers is not in the document]
Comment (kzfsg) [reply src/main/java/seedu/address/model/company/Company.java]: great catch

File: src/main/java/seedu/address/logic/Messages.java [hunk 11]
```diff
@@ -38,22 +38,27 @@ public static String format(Company company) {
         final StringBuilder builder = new StringBuilder();
 
         // Format phone with user-friendly placeholder
-        String phoneDisplay = company.getPhone().value.equals("000")
+        String phoneDisplay = company.getPhone().value == null
                 ? "No phone provided"
                 : company.getPhone().value;
```
Comment (soakedumplings) [inline src/main/java/seedu/address/logic/Messages.java]: I like this implementation of placeholders for phone, emails and address but not for remarks

File: src/main/java/seedu/address/model/company/Company.java (hunk 1 above)
Comment (soakedumplings) [inline src/main/java/seedu/address/model/company/Company.java]: For this javadoc line,
Since we now allow `remark` to be `null`, should we update this to reflect the latest constraints?  
For example:
 Guarantees: name, tags, and status are not null; phone, email, address, and remark can be null.

File: src/main/java/seedu/address/model/company/Remark.java [hunk 12]
```diff
@@ -15,20 +16,20 @@ public class Remark {
     public final String value;
 
     /**
-     * Constructs an {@code Address}.
+     * Constructs a {@code Remark}.
      *
-     * @param remark A valid address.
+     * @param remark A valid remark, or null if no remark is provided.
```
[the remark this answers is not in the document]
Comment (kzfsg) [reply src/main/java/seedu/address/model/company/Remark.java]: A valid remark if non-empty string is provided. Null if no remark or empty remark is provided.

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 13]
```diff
@@ -117,14 +115,7 @@ public void parse_repeatedNonTagValue_failure() {
         assertParseFailure(parser, INVALID_PHONE_DESC + validExpectedCompanyString,
                 Messages.getErrorMessageForDuplicatePrefixes(PREFIX_PHONE));
 
-        // invalid address
-        assertParseFailure(parser, INVALID_ADDRESS_DESC + validExpectedCompanyString,
-                Messages.getErrorMessageForDuplicatePrefixes(PREFIX_ADDRESS));
-
-        // remarks can take any values
-
         // valid value followed by invalid value
```
Comment (soakedumplings) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: I noticed this test is currently being removed. Instead of deleting it entirely, would it be helpful to update it to test an invalid address case based on the `VALIDATION_REGEX`? That way, we still retain coverage for invalid input. 

If there isn’t a regex constraint for address, then removing the test might be reasonable — but I thought I’d check first.

File: src/main/java/seedu/address/storage/JsonAdaptedCompany.java
[the remark this answers is not in the document]
Comment (kzfsg) [reply src/main/java/seedu/address/storage/JsonAdaptedCompany.java]: good catch

**Review by soakedumplings — APPROVED** [review]
Look much better now, just some minor faults here and there. But will merge and create another PR to refine. Good job

File: src/main/java/seedu/address/storage/JsonAdaptedCompany.java
Comment (kzfsg) [reply src/main/java/seedu/address/storage/JsonAdaptedCompany.java]: Good catch

File: src/test/java/seedu/address/storage/JsonAdaptedCompanyTest.java [hunk 14]
```diff
@@ -122,4 +122,33 @@ public void toModelType_nullRemark_throwsIllegalValueException() {
         assertThrows(IllegalValueException.class, expectedMessage, company::toModelType);
     }
 
+    @Test
+    public void toModelType_allNullOptionalFields_success() throws Exception {
+        // Test with all nullable fields (phone, email, address) set to null
+        JsonAdaptedCompany company = new JsonAdaptedCompany(VALID_NAME, null, null, null, VALID_TAGS,
+                VALID_REMARK, VALID_STATUS);
+        assertEquals(null, company.toModelType().getPhone().value);
```
[the remark this answers is not in the document]
Comment (kzfsg) [reply src/test/java/seedu/address/storage/JsonAdaptedCompanyTest.java]: gg idk how i missed this. good catch

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 15]
```diff
@@ -117,8 +116,8 @@ public void parse_repeatedNonTagValue_failure() {
         assertParseFailure(parser, INVALID_PHONE_DESC + validExpectedCompanyString,
                 Messages.getErrorMessageForDuplicatePrefixes(PREFIX_PHONE));
 
-        // invalid address
-        assertParseFailure(parser, INVALID_ADDRESS_DESC + validExpectedCompanyString,
+        // valid address (blank addresses are now allowed, but duplicates are not)
+        assertParseFailure(parser, VALID_ADDRESS_DESC + validExpectedCompanyString,
```
[the remark this answers is not in the document]
Comment (nathantew14) [reply src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: i'm wrong whitespace only is an invalid address

[the remark this answers is not in the document]
Comment (kzfsg) [reply src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: I made the same mistake too. rectified in most recent commit

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java (hunk 6 above)
Comment (nathantew14) [reply src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: my bad, whitespace is invalid address
